File: 이것이 코딩테스트다/06.다이나믹 프로그래밍/3.1로 만들기.py
# 연산을 최소 몇 번을 해서 1로 만들 수 있을까
INF = int(1e9)
n = 26
dp = [0] * (n+1)
dp[1] = 0

for i in range(2, n+1):
  dp[i] = dp[i-1] + 1
  if i % 2 == 0:
    dp[i] = min(dp[i], dp[i // 2] + 1)
  if i % 3 == 0:
    dp[i] = min(dp[i], dp[i // 3] + 1)
  if i % 5 == 0:
    dp[i] = min(dp[i], dp[i // 5] + 1)
print(dp[n])

File: 이것이 코딩테스트다/06.다이나믹 프로그래밍/3.1로 만들기.js
const INF = Number(Infinity);
const n = 26;
let dp = Array(n + 1).fill(INF);

dp[1] = 0;

for (let i = 2; i < n + 1; i++) {
  dp[i] = dp[i - 1] + 1;
  if (i % 2 === 0) {
    dp[i] = Math.min(dp[i], dp[i / 2] + 1);
  }
  if (i % 3 === 0) {
    dp[i] = Math.min(dp[i], dp[i / 3] + 1);
  }
  if (i % 5 === 0) {
    dp[i] = Math.min(dp[i], dp[i / 5] + 1);
  }
}
console.log(dp[n]);
